src/rpc/blockchain.cpp

```diff
@@ -1277,26 +1277,75 @@ void writeBlockChunksAndUpdateMetadata(bool isHexEncoded, HTTPRequest &req,
                                        CBlockIndex& blockIndex, const std::string& rpcReqId,
                                        bool processedInBatch, const RetFormat& rf)
 {
-    auto stream = blockIndex.StreamSyncBlockFromDisk();
-    if (!stream) 
-    {
-        // Block not found on disk. This could be because we have the block
-        // header in our index but don't have the block (for example if a
-        // non-whitelisted node sends us an unrequested long chain of valid
-        // blocks, we add the headers to our index, but don't accept the block).
-        throw block_parse_error(blockIndex.GetBlockHash().GetHex() + " not found on disk");
-    }
-
     CDiskBlockMetaData metadata = blockIndex.GetDiskBlockMetaData();
     bool hasDiskBlockMetaData = !metadata.diskDataHash.IsNull();
+    
+    std::pair<bool, std::string> range_header = req.GetHeader("Range");
+    bool hasRangeHeader = range_header.first;
+
+    uint64_t offset;
+    uint64_t content_length;
+    std::unique_ptr<CForwardReadonlyStream> stream;
+
+    if (!hasDiskBlockMetaData && hasRangeHeader) {
+        stream = blockIndex.StreamSyncBlockFromDisk();
+        do
+        {
+            auto chunk = stream->Read(4096);
+            metadata.diskDataSize += chunk.Size();
+        } while (!stream->EndOfStream());
+
+        blockIndex.SetBlockIndexFileMetaDataIfNotSet(metadata, mapBlockIndex);
```

Comment:
Do we need to do this metadata calculation? If I'm understanding correctly we only need the metadata to report the `Content-Length` and `Contents-Range` back. If we don't have the metadata then couldn't we just skip the `Content-Length` field (as the old code does) and report the `Content-Range` using `/*` (as it says here: https://httpwg.org/specs/rfc9110.html#field.content-range)?